### PR TITLE
implemented sqlite__datediff macro using epoch deltas

### PR DESCRIPTION
**Pull request by tom-juntunen**
- I needed to make some empirical adjustments; added comments for discussion
- I referenced [this](https://til.simonwillison.net/sqlite/unix-timestamp-milliseconds-sqlite) as a resource to handle the partial seconds for datepart in ['second', 'millisecond']. This didn't make enough of a difference to matter at coarser dateparts.
- I looked into using julian days, also referenced in the article above, but they turned out to not be as effective as using epoch time for measuring single millisecond differences, since they ended up being orders of magnitude smaller than the smallest epoch time delta, which makes the empirical values have smaller scale. 

Overall I think it feels a bit experimental and I'd like to do more testing against a full calendar table to see if the empirical factors need refinement, or if this implementation needs to be re-thought.

### CODE REVIEW

File: dbt/include/sqlite/macros/utils/datediff.sql [hunk 1]
```diff
@@ -1,42 +1,89 @@
...
     {% elif datepart == 'week' %}
-        ({{ datediff(first_date, second_date, 'day') }} / 7 + case
-            when strftime('%w', {{first_date}}) <= strftime('%w', {{second_date}}) then
-                case when {{first_date}} <= {{second_date}} then 0 else -1 end
-            else
-                case when {{first_date}} <= {{second_date}} then 1 else 0 end
-        end)
+        CASE
+            WHEN 
+                ((strftime('%s', {{ second_date }}) - strftime('%s', {{ first_date }})) / 604800.0) >= 0.285715
```
Comment (tom-juntunen) [inline dbt/include/sqlite/macros/utils/datediff.sql:34]: Interestingly enough, `1 / 0.285715 = 3.49999`
Which does seem to reflect 7 days of the week if each day contributes 0.5 to that value. 
So this empirical value of 0.28517 seems to reflect the dividing line between two scenarios: when the two dates being measured for their weekly difference are greater than one day apart from each other, or within one day.

If greater than one day apart, we use CEIL to ensure the week difference is 1
If less than one day apart, we use FLOOR to ensure the week difference is 0.

I don't know whether this is a coincidence or if it mathematically makes sense. 

I feel like I need to run more tests, but making a calendar table at the millisecond grain will need to be done carefully to keep the row count low, yet still satisfy additional testing. Or maybe the tests that are setup already are already covering all edge cases. Only thing I can do is try to run more tests that seem different than what we have already and try to break this macro.

Comment (codeforkjeff) [reply dbt/include/sqlite/macros/utils/datediff.sql:34]: If I'm reading it correctly, the implementation for the postgres adapter considers whether the 2nd date "rolls over" into the following week, as determined by whether day of week for second_date is less than that for first_date. So it doesn't have to do with a two-day threshold, it has to do with the fact that 2019-12-31 is a Tuesday, so anything from 2020-01-02 through 2020-01-06 would be considered to have a week difference of 0.

https://github.com/dbt-labs/dbt-postgres/blob/118fbbdc26376dd13cf09b43b31e5eba4235bbe8/dbt/include/postgres/macros/utils/datediff.sql#L11-L17

It's tricky for sqlite because there's no easy way to do a "day of week" calculation, at least with the built-in functions. I wonder if we could find a date library to do this. sqlean, which we're already using, doesn't include any date functions, unfortunately.

Comment (tom-juntunen) [reply dbt/include/sqlite/macros/utils/datediff.sql:34]: > So it doesn't have to do with a two-day threshold, it has to do with the fact that 2019-12-31 is a Tuesday, so anything from 2020-01-02 through 2020-01-06 would be considered to have a week difference of 0.

That makes perfect sense now. Thank you for this explanation! 

Also thanks for linking the dbt-postgres implementation, it's good that we reference these existing implementations because I find the edge cases exemplified by the unit tests aren't capturing everything we need yet, but seeing their implementation helps close the gap and shed light.

Regarding sqlite and "day of week", we don't have access to postgres' `dow`, but we can use sqlite's "%w" string format option: https://www.sqlite.org/lang_datefunc.html

```
select cast(strftime('%w', datetime('now')) as integer);
(1,)
```

Using this we could update our `week` part to something like:

```
{% elif datepart == 'week' %}
        -- Calculate the initial week difference by finding the difference in days and dividing by 7
        (
            ((strftime('%s', {{ second_date }}) - strftime('%s', {{ first_date }})) / 86400) / 7 +
            -- Adjust based on the days of the week to ensure correct week boundaries
            CASE
                WHEN strftime('%w', {{ first_date }}) <= strftime('%w', {{ second_date }})
                THEN CASE
                        WHEN strftime('%s', {{ first_date }}) <= strftime('%s', {{ second_date }})
                        THEN 0
                        ELSE -1
                    END
                ELSE CASE
                        WHEN strftime('%s', {{ first_date }}) <= strftime('%s', {{ second_date }})
                        THEN 1
                        ELSE 0
                    END
            END
        )
```

Though for other parts of the macro, I am now questioning the use of CEIL v FLOOR knowing it likely could go wrong somewhere. I'll look into the postgres native implementation for datediff and see if I can learn a bit more from that to refine this PR.

Comment (tom-juntunen) [reply dbt/include/sqlite/macros/utils/datediff.sql:34]: I also think I should consider using the code that was already adapted from postgresql to the left because it might be more performant in some areas, and also reads more like the postgresql one which would be a common reference point. One example of this is the case statement only being applied on the second operand used for adjustment, not on the whole expression, which avoids calling the same functions (cast and floor/ceil) multiple times on the same expression which is wasteful.
 

For example,
